packages/core/src/render3/di.ts

```diff
@@ -344,6 +344,100 @@ export function injectAttributeImpl(tNode: TNode, attrNameToInject: string): str
   return null;
 }
 
+/**
+ * Returns a `TNode` of the location where the current `LView` is inserted into.
+ *
+ * @param lView an `LView` that we want to find parent `TNode` for.
+ */
+function getParentTNode(lView: LView): TNode|null {
+  const tView = lView[TVIEW];
+  if (tView.type === TViewType.Embedded) {
+    ngDevMode && assertDefined(tView.declTNode, 'Embedded TNodes should have declaration parents.');
+    return tView.declTNode;
+  } else if (tView.type === TViewType.Component) {
+    // Components don't have `TView.declTNode` because each instance of component could be
+    // inserted in different location, hence `TView.declTNode` is meaningless.
+    // Falling back to `T_HOST` in case we cross component boundary.
+    return lView[T_HOST];
+  }
+  // Remaining TNode type is `TViewType.Root` which doesn't have a parent TNode.
+  return null;
+}
+
+/**
+ * Returns the parent TNode and matching LView for a given TNode.
+ *
+ * Either returns the immediate node host or recursively navigates up the tree
+ * until it finds the next valid parent node, if it exists, and returns that.
+ *
+ * @param tNode a `TNode` that we want to find the parent of.
+ * @param lView an `LView` that corresponds to the tNode param.
+ * @param flags an `InjectFlags` that we check to skip @Host.
+ * @returns a tuple of the parent tNode and the corresponding lView of that node or null
+ */
+function getParentNodeView(
+    tNode: TDirectiveHostNode, lView: LView, flags: InjectFlags): [TDirectiveHostNode, LView]|null {
+  ngDevMode && assertTNodeForTView(tNode, lView[TVIEW]);
+  let parentTNode = tNode.parent as TDirectiveHostNode | null;
+  let parentLView = lView;
+
+  // If `tNode.parent` is `null`, it means we are crossing a component boundary.
+  // We should proceed only when there is no `Host` flag provided.
+  if (parentTNode === null && !(flags & InjectFlags.Host)) {
+    parentTNode = getParentTNode(lView) as TDirectiveHostNode | null;
+    if (!parentTNode) return null;
+
+    parentLView = lView[DECLARATION_VIEW]!;
+
+    // In this case, a parent exists and is definitely an element. So it will definitely
+    // an existing lView as the declaration view, which is why we can assume it's defined.
+    ngDevMode && assertDefined(parentLView, 'Declaration LView should be defined');
```

Comment:
I think the comment can be applied to line 390, so we can just group it all together:
```suggestion
    // In this case, a parent exists and is definitely an element. So it will definitely
    // an existing lView as the declaration view, which is why we can assume it's defined.
    parentLView = lView[DECLARATION_VIEW]!;
    ngDevMode && assertDefined(parentLView, 'Declaration LView should be defined');
```

Reply:
Moved.